sling/nlp/embedding/fact-plausibility.cc

```diff
@@ -0,0 +1,547 @@
+
+    // Number of positive/negative correct/wrong predictions.
+    int positive_correct = 0;
+    int positive_wrong = 0;
+    int negative_correct = 0;
+    int negative_wrong = 0;
+  };
+
+  // Evaluate model on positive and negative examples generated from instances.
+  void EvaluateModel(const Handles &instances, Benchmark *benchmark) {
+    static const int batch_size = 2;
+    Random rnd;
+    rnd.seed(seed_);
+    std::vector<Features> premises(batch_size);
+    std::vector<Features> hypotheses(batch_size);
+
+    Instance scorer(flow_.scorer);
+    int *premise = scorer.Get<int>(flow_.premise);
+    int *hypothesis = scorer.Get<int>(flow_.hypothesis);
+    float *logits = scorer.Get<float>(flow_.logits);
+
+    for (int n = 0; n < instances.size(); ++n) {
+      // Get instances for batch.
+      for (int i = 0; i < batch_size; ++i) {
+        int sample = rnd.UniformInt(instances.size());
+        Frame instance(&store_, instances[sample]);
+
+        // Split instance into premise and hypothesis.
+        int num_groups = instance.Get(p_groups_).AsArray().length();
+        int heldout = rnd.UniformInt(num_groups);
+        Split(instance, &premises[i], &hypotheses[i], heldout);
+      }
+
+      // Compute scores for all pairs in batch.
+      for (int i = 0; i < batch_size; ++i) {
+        for (int j = 0; j < batch_size; ++j) {
+          // Check that main hypothesis is not in premise for negative examples.
+          if (i != j && Has(premises[i], hypotheses[j][0])) continue;
+
+          // Set premise and hypothesis for example.
+          Copy(premises[i], premise);
+          Copy(hypotheses[j], hypothesis);
+
+          // Compute plausibility scores.
+          scorer.Compute();
+
+          // Compute accuracy.
+          benchmark->add(i == j, logits[1] > logits[0]);
+        }
+      }
+    }
+  }
+
+  // Split instance into premise and hypothesis.
+  void Split(const Frame &instance,
+             Features *premise, Features *hypothesis,
+             int heldout) {
+    Array facts = instance.Get(p_facts_).AsArray();
+    Array groups = instance.Get(p_groups_).AsArray();
+    int num_groups = groups.length();
+
+    // Add facts to premise, except for one heldout fact group which is
+    // added to the hypothesis.
+    premise->clear();
+    hypothesis->clear();
+    for (int g = 0; g < num_groups; ++g) {
+      // Get range for fact group.
+      int start = g == 0 ? 0 : groups.get(g - 1).AsInt();
+      int end = groups.get(g).AsInt();
+
+      if (g == heldout) {
+        // Add fact group to hyothesis.
+        for (int f = start; f < end; ++f) {
+          hypothesis->push_back(facts.get(f).AsInt());
+        }
+      } else {
+        // Add fact group to premise.
+        for (int f = start; f < end; ++f) {
+          premise->push_back(facts.get(f).AsInt());
+        }
+      }
+    }
+
+    if (premise->size() >= max_features_) {
+      premise->resize(max_features_ - 1);
```

Comment:
why -1? Is it so that we have space to pad the final feature vector with a sentinel -1 at the end? I thought the sentinel was only required for embedding lookup if we were not specifying all max_features entries.

Reply:
As stated above, we always leave room for the terminating -1 in the feature vector.